python/cudf_polars/cudf_polars/dsl/expr.py

```diff
@@ -0,0 +1,1023 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+# TODO: remove need for this
+# ruff: noqa: D101
+"""
+DSL nodes for the polars expression language.
+
+An expression node is a function, `DataFrame -> Column` or `DataFrame -> Scalar`.
+
+The evaluation context is provided by a LogicalPlan node, and can
+affect the evaluation rule as well as providing the dataframe input.
+In particular, the interpretation of the expression language in a
+`GroupBy` node is groupwise, rather than whole frame.
+"""
+
+from __future__ import annotations
+
+import enum
+from enum import IntEnum
+from functools import partial, reduce
+from typing import TYPE_CHECKING, Any, ClassVar, NamedTuple
+
+import pyarrow as pa
+
+from polars.polars import _expr_nodes as pl_expr
+
+import cudf._lib.pylibcudf as plc
+
+from cudf_polars.containers import Column, Scalar
+from cudf_polars.utils import sorting
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+
+    import polars.type_aliases as pl_types
+
+    from cudf_polars.containers import DataFrame
+
+__all__ = [
+    "Expr",
+    "NamedExpr",
+    "Literal",
+    "Col",
+    "BooleanFunction",
+    "StringFunction",
+    "Sort",
+    "SortBy",
+    "Gather",
+    "Filter",
+    "RollingWindow",
+    "GroupedRollingWindow",
+    "Cast",
+    "Agg",
+    "BinOp",
+]
+
+
+class ExecutionContext(IntEnum):
+    FRAME = enum.auto()
+    GROUPBY = enum.auto()
+    ROLLING = enum.auto()
+
+
+class AggInfo(NamedTuple):
+    requests: list[tuple[Expr | None, plc.aggregation.Aggregation, Expr]]
+
+
+class Expr:
+    """
+    An abstract expression object.
+
+    This contains a (potentially empty) tuple of child expressions,
+    along with non-child data. For uniform reconstruction and
+    implementation of hashing and equality schemes, child classes need
+    to provide a certain amount of metadata when they are defined.
+    Specifically, the ``_non_child`` attribute must list, in-order,
+    the names of the slots that are passed to the constructor. The
+    constructor must take arguments in the order ``(*_non_child,
+    *children).``
+    """
+
+    __slots__ = ("dtype", "hash_value", "repr_value")
+    dtype: plc.DataType
+    """Data type of the expression."""
+    hash_value: int
```

Comment:
This is just an internally cached value presumably not intended for public consumption, right?
```suggestion
    _hash_value: int
```